build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt

```diff
@@ -640,3 +594,125 @@ fun <K : Comparable<K>, V> Map<K, V>.dumpMap(dumpValue: (V) -> String): String =
 @TestOnly
 fun <T : Comparable<T>> Collection<T>.dumpCollection(): String =
     "[${sorted().joinToString(", ", transform = Any::toString)}]"
+
+/**
+ * Minimal information about a kotlinc-generated class that will be used to compute recompilation-triggered changes to support incremental
+ * compilation (see [IncrementalJvmCache.saveClassToCache]).
+ *
+ * It's important that this class contain only the minimal required information, as it will be part of the classpath snapshot of the
+ * `KotlinCompile` task and the task needs to support compile avoidance. For example, this class should contain public method signatures,
+ * and should not contain private method signatures, or method implementations.
+ */
+class KotlinClassInfo private constructor(
+    val classId: ClassId,
+    val classKind: KotlinClassHeader.Kind,
+    val classHeaderData: Array<String>?,
+    val classHeaderStrings: Array<String>?,
+    @Suppress("SpellCheckingInspection") val multifileClassName: String?,
+    val constantsMap: LinkedHashMap<String, Any>,
+    val inlineFunctionsMap: LinkedHashMap<String, Long>
+) {
+
+    val className: JvmClassName by lazy { JvmClassName.byClassId(classId) }
+
+    fun scopeFqName(companion: Boolean = false) = when (classKind) {
+        KotlinClassHeader.Kind.CLASS -> {
+            className.fqNameForClassNameWithoutDollars.let { if (companion) it.child(DEFAULT_NAME_FOR_COMPANION_OBJECT) else it }
+        }
+        else -> className.packageFqName
+    }
+
+    companion object {
+
+        fun createFrom(kotlinClass: LocalFileKotlinClass): KotlinClassInfo {
+            return KotlinClassInfo(
+                kotlinClass.classId,
+                kotlinClass.classHeader.kind,
+                kotlinClass.classHeader.data,
+                kotlinClass.classHeader.strings,
+                kotlinClass.classHeader.multifileClassName,
+                getConstantsMap(kotlinClass.fileContents),
+                getInlineFunctionsMap(kotlinClass.classHeader, kotlinClass.fileContents)
+            )
+        }
+
+        /** Creates [KotlinClassInfo] from the given classContents, or returns `null` if the class is not a kotlinc-generated class. */
+        fun tryCreateFrom(classContents: ByteArray): KotlinClassInfo? {
+            return FileBasedKotlinClass.create(classContents) { classId, _, classHeader, _ ->
+                KotlinClassInfo(
+                    classId,
+                    classHeader.kind,
+                    classHeader.data,
+                    classHeader.strings,
+                    classHeader.multifileClassName,
+                    getConstantsMap(classContents),
+                    getInlineFunctionsMap(classHeader, classContents)
+                )
+            }
+        }
+    }
+}
+
+private fun getConstantsMap(bytes: ByteArray): LinkedHashMap<String, Any> {
```

Comment:
nit: make private in the companion class? the same below.

Reply:
I wanted to keep `KotlinClassInfo` simple for readability as it will be used from the Gradle side. I'd like to move these methods to `private inner class ConstantsMap` and `private inner class InlineFunctionsMap` where they originally belonged, but those are inner classes so a companion object is not allowed there.

That said, I can move these to `KotlinClassInfo` if you still think it fits better there.

Reply:
It's ok to leave them as-is.